mmrazor/structures/graph/channel_modules.py

```diff
@@ -0,0 +1,371 @@
+            new_group = group._split_a_new_group(list(range(0, num)))
+            new_groups.append(new_group)
+        return new_groups
+
+    # private methods
+
+    def _clean_channel_info(self, channel: 'ChannelElement', index: int):
+        """Clean the info of a ChannelElement."""
+        self[index].remove(channel)
+
+    def _add_channel_info(self, chanenl: 'ChannelElement', index):
+        """Add the info of a ChannelElemnt."""
+        assert chanenl.group is not self
+        if index not in self.channels:
+            self.channels[index] = []
+        self.channels[index].append(chanenl)
+
+    def _split_a_new_group(self, indexes: List[int]):
+        """Split a part of the group to a new group."""
+        new_group = BaseChannelGroup()
+        j = 0
+        for i in indexes:
+            for channel in copy.copy(self[i]):
+                new_group.add_channel(channel, j)
+            self.channels.pop(i)
+            j += 1
+        self._reindex()
+        return new_group
+
+    def _reindex(self):
+        """Re-index the owning ChannelElements."""
+        j = 0
+        for i in copy.copy(self.channels):
+            if len(self.channels[i]) == 0:
+                self.channels.pop(i)
+            else:
+                if j < i:
+                    for channel in copy.copy(self.channels[i]):
+                        if channel.group is not None:
+                            channel.remove_from_group()
+                        self.add_channel(channel, j)
+                    self.channels.pop(i)
+                    j += 1
+                elif j == i:
+                    pass
+                else:
+                    raise Exception()
+
+    # others
+
+    def __repr__(self) -> str:
+
+        def add_prefix(string: str, prefix='  '):
+            str_list = string.split('\n')
+            str_list = [
+                prefix + line if line != '' else line for line in str_list
+            ]
+            return '\n'.join(str_list)
+
+        def list_repr(lit: List):
+            s = '[\n'
+            for item in lit:
+                s += add_prefix(item.__repr__(), '  ') + '\n'
+            s += ']\n'
+            return s
+
+        s = ('xxxxx_'
+             f'\t{len(self.output_related)},{len(self.input_related)}\n')
+        s += '  output_related:\n'
+        s += add_prefix(list_repr(self.output_related), ' ' * 4)
+        s += '  input_related\n'
+        s += add_prefix(list_repr(self.input_related), ' ' * 4)
+        return s
+
+    def __iter__(self):
+        for i in self.channels:
+            yield i
+
+    def __len__(self):
+        return len(self.channels)
+
+    def __getitem__(self, key):
+        return self.channels[key]
+
+
+class ChannelElement:
+
+    def __init__(self, channel_list: 'ChannelTensor', index: int) -> None:
+        """Each ChannelElement is the basic element of  a ChannelTensor. It
+        records its owing ChannelTensor and BaseChannelGroup.
```

Comment:
It seems that the ChannelEmement does not records its owing ChannelTensor？

Reply:
yes, there is no need to record owing ChannelTensor